metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala

```diff
@@ -1,72 +1,118 @@
 package scala.meta.internal.metals.testProvider
 
-import scala.collection.concurrent.TrieMap
 import scala.collection.mutable
 import scala.concurrent.ExecutionContext
 import scala.concurrent.Future
 
 import scala.meta.internal.metals.BatchedFunction
 import scala.meta.internal.metals.BuildTargets
 import scala.meta.internal.metals.CancelableFuture
+import scala.meta.internal.metals.ClientCommands
 import scala.meta.internal.metals.ClientConfiguration
 import scala.meta.internal.metals.DefinitionProvider
 import scala.meta.internal.metals.MetalsEnrichments._
-import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.clients.language.MetalsLanguageClient
 import scala.meta.internal.metals.debug.BuildTargetClasses
-import scala.meta.internal.metals.debug.BuildTargetClasses.FullyQualifiedClassName
-import scala.meta.internal.metals.debug.BuildTargetClasses.Symbol
+import scala.meta.internal.metals.testProvider.TestExplorerEvent._
+import scala.meta.internal.metals.testProvider.frameworks.JunitTestFinder
+import scala.meta.internal.semanticdb.TextDocuments
+import scala.meta.io.AbsolutePath
 
 import ch.epfl.scala.bsp4j.BuildTarget
-
-private final case class SymbolsPerTarget(
-    target: BuildTarget,
-    testSymbols: TrieMap[Symbol, FullyQualifiedClassName]
-)
+import org.eclipse.lsp4j.Location
 
 final class TestSuitesProvider(
     buildTargets: BuildTargets,
     buildTargetClasses: BuildTargetClasses,
     definitionProvider: DefinitionProvider,
-    clientConfig: ClientConfiguration
+    clientConfig: ClientConfiguration,
+    client: MetalsLanguageClient
 )(implicit ec: ExecutionContext) {
 
-  /**
-   * Cached, already discovered test suites.
-   * Cache is updated via refreshTestSuites method called after compilation is finished.
-   * There is no need to have synchronization since all access is available via
-   * single threaded executor.
-   *
-   * For every test suite a.MunitTestSuite BSP returns 2 symbols. However, only one of them
-   * has location in source code and it's useful.
-   * That's why FullyQualifiedClassName is used as a key.
-   * "a/MunitTestSuite., a.MunitTestSuite"
-   * "a/MunitTestSuite#, a.MunitTestSuite"
-   */
-  private val cachedTestSuites =
-    mutable.Map[FullyQualifiedClassName, TestEntry]()
+  private val index = new TestSuitesIndex
+  private val junitTestFinder = new JunitTestFinder
 
   val refreshTestSuites = new BatchedFunction[Unit, Unit](_ =>
     CancelableFuture(doRefreshTestSuites())
   )
 
+  def onChange(docs: TextDocuments, file: AbsolutePath): Unit = {
```

Comment:
So thinking about the flow here a bit, I wonder if mimicking what we do with TVP might be a good idea. Mainly if a user doesn't have the test view open, then they are still going to be sent the `update-test-explorer` events. Do you think it'd make sense to keep track of the state of the explorer window like we do with TVP and if this is closed then don't even sent the update events? Then when they navigate back to it do a `metals.discover-test-suites`?

Reply:
Thinking more about this, I'm not sure what would be best because in the scenario where we need to use `metals.discover-test-suites` right now we are sending everything, which is fine, but I'm glad we are taking the approach withe `update-test-explorer` to only send what is updated. That makes it 1000% time easier to manage on the client side imo. For example VS Code does some magic behind the scenes in TVP where it without any interruptions nicely updates the parts of the tree that are updated and only those parts. However in `nvim-metals` it isn't quite that smooth because we are just notified that a build target is updated, and then we replace the entire thing. It's not ideal, but if not I'd need to manually compare the two trees and only update the parts that have changed while retaining the collapse state etc. It's a lot of manual stuff that is free in VS Code.

All that to say, I'm glad we are doing it this way for the explorer and I'm just trying to think of a way to not fire a lot events to the client when it's not needed, but maybe it doesn't matter. wdyt?

Reply:
From vscode's perspective, Metals have to fire events when:
- file which contains tests is opened
- TestExplorer view is opened
Otherwise, users can see stale information (gutter icons in file and entries is TE view). 
Sending events only for already "visited" test suites should be sufficient to keep a valid state. By "visited" I mean every test suite which has testcases listed in TE view.

A short GIF to show what I meant.

![my-app-1642959912310](https://user-images.githubusercontent.com/37124721/150691044-d7200d62-3263-43cb-bfe3-84047ca44321.gif)

If I understood correctly you're concerned about unwanted events, yes? A few strategies like: `send all`, `send diffs` can be included in logic and one will be chosen depending on the client. Then vscode will get all events (because it needs them) and `nvim` can opt into `send all` and don't receive unnecessary events.

Reply:
Actually thinking about it more since I wrote this, I don't think it's a valid worry. I was worried that it'd be like TVP where _every_ time anything in the target changes you get sent stuff that changes were made, which can then be really expensive if you need to recreate the state of that tree. However since we are only ever sending diffs, I don't think it's even a big deal. Just just ignore that concern.